README.md
[![pub package](https://img.shields.io/badge/pub-v0.0.6-blue)](https://pub.dev/packages/analog_clock_picker)

A customable time picker using analog clock format

## Features

- Customizable background color and widget
- Customizable clock hand color and widget
- Ability to change period AM/PM type via Controller
- Fluid animation movement of clock hand

## Screenshot

<p float="left">
    <img src="https://raw.githubusercontent.com/yasfdany/analog_clock_picker/master/doc/screenshots/ss_minimal.jpg" width="200px">
    <img src="https://raw.githubusercontent.com/yasfdany/analog_clock_picker/master/doc/screenshots/ss_custom_bg.jpg" width="200px">
    <img src="https://raw.githubusercontent.com/yasfdany/analog_clock_picker/master/doc/screenshots/ss_custom_ring.jpg" width="200px">
</p>

## Demo Video

<p float="left">
  <img src="https://raw.githubusercontent.com/yasfdany/analog_clock_picker/master/doc/gif/demo.gif" width="200px">
</p>

## Getting started

Add package to `pubspec.yaml`

```yaml
dependencies:
    analog_clock_picker: 0.0.6
```

## How to use

Create controller for control the value of the clock

```dart
AnalogClockController analogClockController = AnalogClockController();
```

You can provide default value inside the controller

```dart
AnalogClockController analogClockController = AnalogClockController(
  value: DateTime.now(),
  periodType: PeriodType.am,
  onPeriodTypeChange: (date, period) {
    //TODO : Do Something
  },
);
```

you can set period change listener after initialize it

```dart
analogClockController.setOnPeriodTypeChangeListener((date, period) {
      //TODO : Do Something
});
```

Minimal usage

```dart
AnalogClockPicker(
  controller: analogClockController,
  size: MediaQuery.of(context).size.width * 0.74,
  secondHandleColor: Colors.red,
  minutesHandleColor: Colors.black,
  hourHandleColor: Colors.black,
  clockBackground: Image.asset(
    AssetImages.clockBackground,
  ),
  onClockChange: (date){
    //TODO : Do Something
  },
)
```

CHANGELOG.md
## 0.0.6

[0.0.1]

Initial Release

[0.0.2]

Fix minor bug

[0.0.3]

Add ability to custom ring widget

[0.0.6]

- Add default widget for clock background
- Fix scaling problem in clock components

example/lib/main.dart
import 'package:analog_clock_picker/analog_clock_picker.dart';
import 'package:flutter/material.dart';
import 'package:intl/intl.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key}) : super(key: key);

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  AnalogClockController analogClockController = AnalogClockController();

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text(
          "Analog clock picker",
        ),
      ),
      body: SizedBox(
        width: double.infinity,
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.center,
          mainAxisAlignment: MainAxisAlignment.center,
          children: [
            AnalogClockPicker(
              clockBackground: Image.asset(
                "assets/images/clock_background.png",
              ),
              clockBackgroundColor: Colors.white,
              secondHandleColor: Colors.red,
              minutesHandleColor: Colors.black,
              hourHandleColor: Colors.black,
              controller: analogClockController,
              size: MediaQuery.of(context).size.width * 0.68,
            ),
            ValueListenableBuilder<DateTime>(
              valueListenable: analogClockController,
              builder: (context, value, _) {
                return Container(
                  margin: const EdgeInsets.only(top: 42),
                  child: Row(
                    mainAxisAlignment: MainAxisAlignment.center,
                    children: [
                      Container(
                        margin: const EdgeInsets.only(right: 12),
                        child: Text(
                          DateFormat("HH:mm").format(value),
                          style: const TextStyle(
                            fontSize: 40,
                            color: Colors.black,
                            fontWeight: FontWeight.bold,
                          ),
                        ),
                      ),
                      ElevatedButton(
                        onPressed: () {
                          analogClockController.setPeriodType(
                            analogClockController.periodType == PeriodType.am
                                ? PeriodType.pm
                                : PeriodType.am,
                          );
                        },
                        child: Text(
                          analogClockController.periodType == PeriodType.am
                              ? "AM"
                              : "PM",
                          style: const TextStyle(
                            fontSize: 18,
                          ),
                        ),
                      ),
                    ],
                  ),
                );
              },
            ),
          ],
        ),
      ),
    );
  }
}


lib/hour_handle_widget.dart
import 'dart:math';

import 'package:flutter/material.dart';

import 'analog_clock_picker.dart';

class HourHandleWidget extends StatelessWidget {
  final Color handleColor;
  final double handleHeight;
  final AnimationController controller;
  final Function(int hour) onDragHandleEnd;
  final Offset centerPosition;
  final Widget? handleWidget;
  final double radius;

  const HourHandleWidget({
    Key? key,
    required this.handleColor,
    required this.handleHeight,
    required this.controller,
    required this.onDragHandleEnd,
    required this.centerPosition,
    required this.radius,
    this.handleWidget,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return AnimatedBuilder(
      animation: controller,
      builder: (context, value) => Transform.rotate(
        angle: controller.value / 180 * pi,
        child: GestureDetector(
          onPanEnd: (end) {
            stickyHourHand();
          },
          onPanUpdate: (pan) {
            controller.value = calculateAngle(
              centerPosition,
              pan.globalPosition,
            );
            onDragHandleEnd(((controller.value / 360) * 12).toInt());
          },
          child: Container(
            margin: EdgeInsets.only(bottom: handleHeight),
            child: Container(
              //transparent container with horizontal margin for expand handle touch area
              color: Colors.transparent,
              padding: const EdgeInsets.symmetric(horizontal: 18),
              child: handleWidget != null
                  ? SizedBox(
                      height: handleHeight,
                      child: handleWidget,
                    )
                  : Container(
                      decoration: BoxDecoration(
                        color: handleColor,
                        borderRadius: BorderRadius.circular(12),
                      ),
                      width: radius * 0.068,
                      height: handleHeight,
                    ),
            ),
          ),
        ),
      ),
    );
  }

  void setTime() {
    if (controller.value == 360) {
      onDragHandleEnd(0);
    } else {
      onDragHandleEnd(((controller.value / 360) * 12).toInt());
    }
  }

  //Logic for sticky hour handle
  void stickyHourHand() {
    //search angle beetween all possible angles
    for (int i = 0; i < hourAngles.length; i++) {
      if (i < hourAngles.length - 1) {
        double angle1 = hourAngles[i];
        double angle2 = hourAngles[i + 1];

        if (controller.value > angle1 && controller.value < angle2) {
          //calculate current hand angle distance beetween previous and next possible angle
          double angleDistance1 = (controller.value - angle1).abs();
          double angleDistance2 = (controller.value - angle2).abs();

          //assign the closest possible angle from current hand angle
          if (angleDistance1 > angleDistance2) {
            controller
                .animateTo(angle2, duration: const Duration(milliseconds: 100))
                .whenComplete(setTime);
          } else {
            controller
                .animateTo(angle1, duration: const Duration(milliseconds: 100))
                .whenComplete(setTime);
          }

          break;
        }
      }
    }
  }
}


lib/minutes_handle_widget.dart
import 'dart:math';

import 'package:flutter/material.dart';

import 'analog_clock_picker.dart';

class MinutesHandleWidget extends StatelessWidget {
  final Color handleColor;
  final double handleHeight;
  final AnimationController controller;
  final Function(int minute) onDragHandleEnd;
  final Offset centerPosition;
  final Widget? handleWidget;
  final double radius;

  const MinutesHandleWidget({
    Key? key,
    required this.handleColor,
    required this.handleHeight,
    required this.controller,
    required this.onDragHandleEnd,
    required this.centerPosition,
    required this.radius,
    this.handleWidget,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return AnimatedBuilder(
      animation: controller,
      builder: (context, value) => Transform.rotate(
        angle: controller.value / 180 * pi,
        child: GestureDetector(
          onPanEnd: (end) {
            stickyMinuteHand();
          },
          onPanUpdate: (pan) {
            controller.value = calculateAngle(
              centerPosition,
              pan.globalPosition,
            );

            onDragHandleEnd(((controller.value / 360) * 60).toInt());
          },
          child: Container(
            margin: EdgeInsets.only(bottom: handleHeight),
            child: Container(
              //transparent container with horizontal margin for expand handle touch area
              color: Colors.transparent,
              padding: const EdgeInsets.symmetric(horizontal: 18),
              child: handleWidget != null
                  ? SizedBox(
                      height: handleHeight,
                      child: handleWidget,
                    )
                  : Container(
                      decoration: BoxDecoration(
                        color: handleColor,
                        borderRadius: BorderRadius.circular(12),
                      ),
                      width: radius * 0.058,
                      height: handleHeight,
                    ),
            ),
          ),
        ),
      ),
    );
  }

  //Logic for sticky minute handle
  void stickyMinuteHand() {
    //search angle beetween all possible angles
    for (int i = 0; i < minutesAngles.length; i++) {
      if (i < minutesAngles.length - 1) {
        double angle1 = minutesAngles[i];
        double angle2 = minutesAngles[i + 1];

        if (controller.value > angle1 && controller.value < angle2) {
          //calculate current hand angle distance beetween previous and next possible angle
          double angleDistance1 = (controller.value - angle1).abs();
          double angleDistance2 = (controller.value - angle2).abs();

          //assign the closest possible angle from current hand angle
          if (angleDistance1 > angleDistance2) {
            controller
                .animateTo(angle2, duration: const Duration(milliseconds: 100))
                .whenComplete(() {
              onDragHandleEnd(((controller.value / 360) * 60).toInt());
            });
          } else {
            controller
                .animateTo(angle1, duration: const Duration(milliseconds: 100))
                .whenComplete(() {
              onDragHandleEnd(((controller.value / 360) * 60).toInt());
            });
          }

          break;
        }
      }
    }
  }
}


lib/clock_background_widget.dart
// ignore_for_file: prefer_const_constructors

import 'dart:math';

import 'package:analog_clock_picker/analog_clock_picker.dart';
import 'package:flutter/material.dart';

class ClockBackgroundWidget extends StatelessWidget {
  final double radius;
  final TextStyle? textStyle;
  final Color? minuteStickColor;
  final Color? hourStickColor;

  const ClockBackgroundWidget({
    Key? key,
    required this.radius,
    this.textStyle,
    this.minuteStickColor = Colors.black,
    this.hourStickColor = Colors.black,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Stack(
      children: [
        for (double angle in minutesAngles.sublist(1, minutesAngles.length))
          Transform.rotate(
            angle: angle * pi / 180,
            child: Container(
              alignment: Alignment.center,
              child: Column(
                crossAxisAlignment: CrossAxisAlignment.center,
                children: [
                  Container(
                    margin: EdgeInsets.symmetric(
                      vertical: radius * 0.024,
                    ),
                    width: 2,
                    height: radius * 0.04,
                    color: minuteStickColor,
                  ),
                ],
              ),
            ),
          ),
        for (double angle in hourAngles.sublist(1, hourAngles.length))
          Transform.rotate(
            angle: angle * pi / 180,
            child: Container(
              alignment: Alignment.center,
              child: Column(
                crossAxisAlignment: CrossAxisAlignment.center,
                children: [
                  Container(
                    margin: EdgeInsets.symmetric(
                      vertical: radius * 0.024,
                    ),
                    width: 4,
                    height: radius * 0.08,
                    color: hourStickColor,
                  ),
                  Text(
                    "${angle ~/ 30}",
                    style: textStyle ??
                        const TextStyle(
                          fontSize: 18,
                        ),
                  ),
                ],
              ),
            ),
          )
      ],
    );
  }
}


lib/analog_clock_picker.dart
library analog_clock_picker;

import 'dart:math';

import 'package:analog_clock_picker/clock_background_widget.dart';
import 'package:analog_clock_picker/hour_handle_widget.dart';
import 'package:analog_clock_picker/minutes_handle_widget.dart';
import 'package:analog_clock_picker/second_handle_widget.dart';
import 'package:flutter/material.dart';
import 'package:flutter/scheduler.dart';

//generate minutes angle
final List<double> minutesAngles = [for (int i = 0; i <= 60; i++) i * 6];
//generate hour angle
final List<double> hourAngles = [for (int i = 0; i <= 12; i++) i * 30];

//Formula for calculating angle from two points
double calculateAngle(Offset origin, Offset target) {
  var dx = origin.dx - target.dx;
  var dy = origin.dy - target.dy;

  var theta = atan2(-dy, -dx);
  theta *= 180 / pi;
  theta += 90;
  if (theta < 0) theta += 360;

  return theta;
}

enum PeriodType {
  am,
  pm,
}

class AnalogClockController extends ValueNotifier<DateTime> {
  PeriodType? _periodType;
  PeriodType? get periodType => _periodType;
  Function(DateTime dateTime, PeriodType periodType)? onPeriodTypeChange;

  void setOnPeriodTypeChangeListener(
    Function(DateTime dateTime, PeriodType periodType) listener,
  ) {
    onPeriodTypeChange = listener;
  }

  void setPeriodType(PeriodType periodType) {
    _periodType = periodType;
    DateTime newDate = DateTime(
      value.year,
      value.month,
      value.day,
      periodType == PeriodType.am
          ? (value.hour > 12 ? value.hour - 12 : value.hour)
          : (value.hour > 12 ? value.hour : value.hour + 12),
      value.minute,
      0,
    );
    value = newDate;
    if (onPeriodTypeChange != null) onPeriodTypeChange!(newDate, periodType);
  }

  AnalogClockController({
    DateTime? value,
    PeriodType? periodType = PeriodType.am,
    this.onPeriodTypeChange,
  }) : super(value ?? DateTime.now()) {
    _periodType = periodType;
  }
}

class AnalogClockPicker extends StatefulWidget {
  final double size;
  final Color clockBackgroundColor;
  final Color secondHandleColor;
  final Color minutesHandleColor;
  final Color hourHandleColor;
  final Widget? clockBackground;
  final Widget? secondHandleWidget;
  final Widget? minutesHandleWidget;
  final Widget? hourHandleWidget;
  final Widget? ringWidget;
  final AnalogClockController controller;
  final Function(DateTime dateTime)? onClockChange;
  final TextStyle? clockTextStyle;
  final Color? minuteStickColor;
  final Color? hourStickColor;

  const AnalogClockPicker({
    Key? key,
    required this.controller,
    required this.size,
    this.onClockChange,
    this.secondHandleColor = Colors.white,
    this.minutesHandleColor = Colors.white,
    this.hourHandleColor = Colors.white,
    this.clockBackgroundColor = Colors.white,
    this.clockBackground,
    this.secondHandleWidget,
    this.minutesHandleWidget,
    this.hourHandleWidget,
    this.ringWidget,
    this.clockTextStyle,
    this.minuteStickColor = Colors.black,
    this.hourStickColor = Colors.black,
  }) : super(key: key);

  @override
  State<AnalogClockPicker> createState() => _AnalogClockPickerState();
}

class _AnalogClockPickerState extends State<AnalogClockPicker>
    with TickerProviderStateMixin {
  GlobalKey key = GlobalKey();

  late AnimationController hourAngle;
  late AnimationController minutesAngle;
  Offset? centerPosition;
  double radius = 0;
  RenderBox? box;

  @override
  void initState() {
    super.initState();
    if (widget.controller.value.hour > 12) {
      widget.controller.setPeriodType(PeriodType.pm);
    }
    hourAngle = AnimationController(
      vsync: this,
      lowerBound: 0,
      upperBound: 360,
    );
    minutesAngle = AnimationController(
      vsync: this,
      lowerBound: 0,
      upperBound: 360,
    );
    if (widget.controller.value.hour > 12) {
      hourAngle.value = (widget.controller.value.hour - 12) * 30;
    } else {
      hourAngle.value = widget.controller.value.hour * 30;
    }
    minutesAngle.value = widget.controller.value.minute * 6;

    radius = widget.size / 2;

    SchedulerBinding.instance?.addPostFrameCallback((_) async {
      if (key.currentContext != null) {
        box = key.currentContext?.findRenderObject() as RenderBox;
      }
      Offset? position = box?.localToGlobal(Offset.zero);

      setState(() {
        centerPosition = Offset(
          (position?.dx ?? 0) + (widget.size / 2),
          (position?.dy ?? 0) + (widget.size / 2),
        );
      });
    });
  }

  @override
  Widget build(BuildContext context) {
    radius = widget.size / 2;

    return SizedBox(
      key: key,
      width: widget.size,
      height: widget.size,
      child: Stack(
        alignment: Alignment.center,
        children: [
          Container(
            decoration: BoxDecoration(
              shape: BoxShape.circle,
              color: widget.clockBackgroundColor,
            ),
            width: widget.size + 10,
            height: widget.size + 10,
          ),
          widget.clockBackground ??
              ClockBackgroundWidget(
                radius: radius,
                textStyle: widget.clockTextStyle,
                hourStickColor: widget.hourStickColor,
                minuteStickColor: widget.minuteStickColor,
              ),
          SecondHandleWidget(
            handleHeight: radius * 0.68,
            handleColor: widget.secondHandleColor,
            handleWidget: widget.secondHandleWidget,
          ),
          HourHandleWidget(
            handleColor: widget.hourHandleColor,
            handleHeight: radius * 0.56,
            controller: hourAngle,
            onDragHandleEnd: (hour) {
              widget.controller.value = DateTime(
                widget.controller.value.year,
                widget.controller.value.month,
                widget.controller.value.day,
                widget.controller.periodType == PeriodType.am
                    ? hour
                    : hour + 12,
                widget.controller.value.minute,
                0,
              );
              if (widget.onClockChange != null) {
                widget.onClockChange!(widget.controller.value);
              }
            },
            centerPosition: centerPosition ?? Offset.zero,
            handleWidget: widget.hourHandleWidget,
            radius: radius,
          ),
          MinutesHandleWidget(
            handleColor: widget.minutesHandleColor,
            handleHeight: radius * 0.68,
            controller: minutesAngle,
            onDragHandleEnd: (minute) {
              widget.controller.value = DateTime(
                widget.controller.value.year,
                widget.controller.value.month,
                widget.controller.value.day,
                widget.controller.value.hour,
                minute,
                0,
              );
              if (widget.onClockChange != null) {
                widget.onClockChange!(widget.controller.value);
              }
            },
            centerPosition: centerPosition ?? Offset.zero,
            handleWidget: widget.minutesHandleWidget,
            radius: radius,
          ),
          widget.ringWidget != null
              ? SizedBox(
                  width: radius * 0.16,
                  height: radius * 0.16,
                  child: widget.ringWidget,
                )
              : Container(
                  width: radius * 0.16,
                  height: radius * 0.16,
                  decoration: BoxDecoration(
                    shape: BoxShape.circle,
                    color: Colors.white,
                    border: Border.all(
                      color: Colors.black,
                      width: radius * 0.04,
                    ),
                  ),
                ),
        ],
      ),
    );
  }
}
